File: macOS/macAI/Data/AnalysisProviders.swift
import Foundation
import os

/// Format-specific API implementations for analysis (reports).
extension AnalysisRepository {

    // MARK: - OpenAI Compatible (28 providers)

    func analyzeWithOpenAiCompatible(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters,
        customBaseUrl: String? = nil
    ) async -> AnalysisResponse {
        // Check if model uses Responses API
        if usesResponsesApi(service, model: model) {
            return await analyzeWithResponsesApi(service: service, apiKey: apiKey, prompt: prompt, model: model, params: params, customBaseUrl: customBaseUrl)
        }

        let messages = buildOpenAiMessages(prompt: prompt, systemPrompt: params.systemPrompt)

        let request = OpenAiRequest(
            model: model,
            messages: messages,
            max_tokens: params.maxTokens,
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            frequency_penalty: params.frequencyPenalty,
            presence_penalty: params.presencePenalty,
            stop: params.stopSequences,
            seed: service.seedFieldName == "seed" ? params.seed : nil,
            random_seed: service.seedFieldName == "random_seed" ? params.seed : nil,
            response_format: params.responseFormatJson ? OpenAiResponseFormat(type: "json_object") : nil,
            return_citations: service.supportsCitations ? params.returnCitations : nil,
            search_recency_filter: service.supportsSearchRecency ? params.searchRecency : nil,
            search: params.searchEnabled ? true : nil
        )

        let url = ApiClient.chatUrl(for: service, customBaseUrl: customBaseUrl)

        do {
            let (response, httpResponse): (OpenAiResponse, HTTPURLResponse) = try await ApiClient.shared.request(
                url: url, body: request, service: service, apiKey: apiKey
            )

            if let error = response.error {
                return AnalysisResponse(service: service, error: error.message ?? "Unknown error",
                                       httpStatusCode: httpResponse.statusCode)
            }

            let content = response.choices?.first?.message.content
                ?? response.choices?.first?.message.reasoning_content

            let usage = response.usage
            let tokenUsage = TokenUsage(
                inputTokens: usage?.effectiveInputTokens ?? 0,
                outputTokens: usage?.effectiveOutputTokens ?? 0,
                apiCost: Self.extractApiCost(usage: usage, provider: service)
            )

            return AnalysisResponse(
                service: service,
                analysis: content,
                error: content == nil ? "No content in response" : nil,
                tokenUsage: tokenUsage,
                citations: response.citations,
                searchResults: response.search_results,
                relatedQuestions: response.related_questions,
                rawUsageJson: formatUsageJson(usage),
                httpStatusCode: httpResponse.statusCode
            )
        } catch {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        }
    }

    // MARK: - OpenAI Responses API

    private func analyzeWithResponsesApi(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters,
        customBaseUrl: String? = nil
    ) async -> AnalysisResponse {
        let request = OpenAiResponsesRequest(model: model, input: prompt, instructions: params.systemPrompt)

        let base = customBaseUrl ?? service.baseUrl
        let normalizedBase = base.hasSuffix("/") ? base : "\(base)/"
        let url = "\(normalizedBase)v1/responses"

        do {
            let (response, httpResponse): (OpenAiResponsesApiResponse, HTTPURLResponse) = try await ApiClient.shared.request(
                url: url, body: request, service: service, apiKey: apiKey
            )

            if let error = response.error {
                return AnalysisResponse(service: service, error: error.message ?? "Unknown error",
                                       httpStatusCode: httpResponse.statusCode)
            }

            // Extract content from output blocks
            let content = response.output?
                .compactMap { msg in msg.content?.first(where: { $0.type == "output_text" })?.text ?? msg.content?.first?.text }
                .first

            let usage = response.usage
            let tokenUsage = TokenUsage(
                inputTokens: usage?.effectiveInputTokens ?? 0,
                outputTokens: usage?.effectiveOutputTokens ?? 0,
                apiCost: Self.extractApiCost(usage: usage, provider: service)
            )

            return AnalysisResponse(
                service: service,
                analysis: content,
                error: content == nil ? "No content in response" : nil,
                tokenUsage: tokenUsage,
                rawUsageJson: formatUsageJson(usage),
                httpStatusCode: httpResponse.statusCode
            )
        } catch {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        }
    }

    // MARK: - Anthropic Claude

    func analyzeWithClaude(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters
    ) async -> AnalysisResponse {
        let request = ClaudeRequest(
            model: model,
            max_tokens: params.maxTokens ?? 4096,
            messages: [ClaudeMessage(role: "user", content: prompt)],
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            system: params.systemPrompt,
            stop_sequences: params.stopSequences,
            search: params.searchEnabled ? true : nil
        )

        let base = service.baseUrl.hasSuffix("/") ? service.baseUrl : "\(service.baseUrl)/"
        let url = "\(base)v1/messages"

        do {
            let (response, httpResponse): (ClaudeResponse, HTTPURLResponse) = try await ApiClient.shared.request(
                url: url, body: request, service: service, apiKey: apiKey
            )

            if let error = response.error {
                return AnalysisResponse(service: service, error: error.message ?? "Unknown error",
                                       httpStatusCode: httpResponse.statusCode)
            }

            let content = response.content?.first(where: { $0.type == "text" })?.text

            let tokenUsage = TokenUsage(
                inputTokens: response.usage?.input_tokens ?? 0,
                outputTokens: response.usage?.output_tokens ?? 0,
                apiCost: Self.extractApiCost(usage: response.usage)
            )

            return AnalysisResponse(
                service: service,
                analysis: content,
                error: content == nil ? "No content in response" : nil,
                tokenUsage: tokenUsage,
                rawUsageJson: formatUsageJson(response.usage),
                httpStatusCode: httpResponse.statusCode
            )
        } catch {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        }
    }

    // MARK: - Google Gemini

    func analyzeWithGemini(
        service: AppService,
        apiKey: String,
        prompt: String,
        model: String,
        params: AgentParameters
    ) async -> AnalysisResponse {
        let config = GeminiGenerationConfig(
            temperature: params.temperature,
            topP: params.topP,
            topK: params.topK,
            maxOutputTokens: params.maxTokens,
            stopSequences: params.stopSequences
        )

        let systemInstruction = params.systemPrompt.map {
            GeminiContent(parts: [GeminiPart(text: $0)])
        }

        let request = GeminiRequest(
            contents: [GeminiContent(parts: [GeminiPart(text: prompt)], role: "user")],
            generationConfig: config,
            systemInstruction: systemInstruction
        )

        let url = ApiClient.geminiGenerateUrl(baseUrl: service.baseUrl, model: model, apiKey: apiKey)

        do {
            // Gemini uses a dummy service for auth (query param handled in URL)
            let dummyService = AppService(id: "gemini_temp", displayName: "", baseUrl: "", defaultModel: "", apiFormat: .google)
            let (response, httpResponse): (GeminiResponse, HTTPURLResponse) = try await ApiClient.shared.request(
                url: url, body: request, service: dummyService, apiKey: apiKey
            )

            if let error = response.error {
                return AnalysisResponse(service: service, error: error.message ?? "Unknown error",
                                       httpStatusCode: httpResponse.statusCode)
            }

            let content = response.candidates?.first?.content?.parts.first?.text

            let usage = response.usageMetadata
            let tokenUsage = TokenUsage(
                inputTokens: usage?.promptTokenCount ?? 0,
                outputTokens: usage?.candidatesTokenCount ?? 0,
                apiCost: Self.extractApiCost(usage: usage)
            )

            return AnalysisResponse(
                service: service,
                analysis: content,
                error: content == nil ? "No content in response" : nil,
                tokenUsage: tokenUsage,
                rawUsageJson: formatUsageJson(usage),
                httpStatusCode: httpResponse.statusCode
            )
        } catch {
            return AnalysisResponse(service: service, error: error.localizedDescription)
        }
    }

    // MARK: - Helpers

    private func buildOpenAiMessages(prompt: String, systemPrompt: String?) -> [OpenAiMessage] {
        var messages: [OpenAiMessage] = []
        if let sys = systemPrompt, !sys.isEmpty {
            messages.append(OpenAiMessage(role: "system", content: sys))
        }
        messages.append(OpenAiMessage(role: "user", content: prompt))
        return messages
    }

    // MARK: - Agent Analysis Entry Point

    func analyzeWithAgent(
        agent: Agent,
        content: String,
        prompt: String,
        settings: Settings,
        overrideParams: AgentParameters? = nil
    ) async -> AnalysisResponse {
        guard let provider = agent.provider else {
            return AnalysisResponse(service: AppService.entries.first!, error: "Provider not found for agent \(agent.name)")
        }

        let effectiveApiKey = settings.getEffectiveApiKeyForAgent(agent)
        guard !effectiveApiKey.isEmpty else {
            return AnalysisResponse(service: provider, error: "API key not configured for agent \(agent.name)", agentName: agent.name)
        }

        let finalPrompt = buildPrompt(prompt, content: content, agent: agent)
        let agentResolvedParams = settings.resolveAgentParameters(agent)
        let params = validateParams(mergeParameters(agentResolvedParams, overrideParams))
        let customBaseUrl = settings.getEffectiveEndpointUrlForAgent(agent)
        let effectiveModel = settings.getEffectiveModelForAgent(agent)

        let makeApiCall: () async throws -> AnalysisResponse = {
            var result: AnalysisResponse
            switch provider.apiFormat {
            case .anthropic:
                result = await self.analyzeWithClaude(service: provider, apiKey: effectiveApiKey, prompt: finalPrompt, model: effectiveModel, params: params)
            case .google:
                result = await self.analyzeWithGemini(service: provider, apiKey: effectiveApiKey, prompt: finalPrompt, model: effectiveModel, params: params)
            case .openaiCompatible:
                result = await self.analyzeWithOpenAiCompatible(service: provider, apiKey: effectiveApiKey, prompt: finalPrompt, model: effectiveModel, params: params, customBaseUrl: customBaseUrl)
            }
            result.agentName = agent.name
            result.promptUsed = finalPrompt
            return result
        }

        return await withRetry(
            label: "Agent \(agent.name)",
            makeCall: makeApiCall,
            isSuccess: { $0.isSuccess },
            errorResult: { error in
                AnalysisResponse(service: provider, error: "Network error after retry: \(error.localizedDescription)", agentName: agent.name)
            }
        )
    }
}

File: windows/AI/Data/AnalysisProviders.cs
using System.Text.Json;
using AI.ViewModels;

namespace AI.Data;

public static class AnalysisProviders
{
    private static readonly AnalysisRepository Repo = AnalysisRepository.Instance;

    public static async Task<AnalysisResponse> AnalyzeWithAgent(
        ViewModels.Agent agent, string content, string prompt,
        ViewModels.Settings settings, AgentParameters? overrideParams = null)
    {
        var provider = agent.Provider;
        if (provider == null)
            return new AnalysisResponse(AppService.Entries.First(), error: $"Provider not found for agent {agent.Name}");

        var effectiveApiKey = settings.GetEffectiveApiKeyForAgent(agent);
        if (string.IsNullOrEmpty(effectiveApiKey))
            return new AnalysisResponse(provider, error: $"API key not configured for agent {agent.Name}", agentName: agent.Name);

        var finalPrompt = Repo.BuildPrompt(prompt, content, agent);
        var agentResolvedParams = settings.ResolveAgentParameters(agent);
        var mergedParams = Repo.ValidateParams(Repo.MergeParameters(agentResolvedParams, overrideParams));
        var customBaseUrl = settings.GetEffectiveEndpointUrlForAgent(agent);
        var effectiveModel = settings.GetEffectiveModelForAgent(agent);

        return await Repo.WithRetry(
            $"Agent {agent.Name}",
            async () =>
            {
                var result = provider.ApiFormat switch
                {
                    ApiFormat.Anthropic => await AnalyzeWithClaude(provider, effectiveApiKey, finalPrompt, effectiveModel, mergedParams),
                    ApiFormat.Google => await AnalyzeWithGemini(provider, effectiveApiKey, finalPrompt, effectiveModel, mergedParams),
                    _ => await AnalyzeWithOpenAiCompatible(provider, effectiveApiKey, finalPrompt, effectiveModel, mergedParams, customBaseUrl)
                };
                result.AgentName = agent.Name;
                result.PromptUsed = finalPrompt;
                return result;
            },
            r => r.IsSuccess,
            ex => new AnalysisResponse(provider, error: $"Network error after retry: {ex.Message}", agentName: agent.Name)
        );
    }

    public static async Task<AnalysisResponse> AnalyzeWithOpenAiCompatible(
        AppService service, string apiKey, string prompt, string model,
        AgentParameters parms, string? customBaseUrl = null)
    {
        if (Repo.UsesResponsesApi(service, model))
            return await AnalyzeWithResponsesApi(service, apiKey, prompt, model, parms, customBaseUrl);

        var messages = BuildOpenAiMessages(prompt, parms.SystemPrompt);
        var request = new OpenAiRequest
        {
            Model = model,
            Messages = messages,
            MaxTokens = parms.MaxTokens,
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            FrequencyPenalty = parms.FrequencyPenalty,
            PresencePenalty = parms.PresencePenalty,
            Stop = parms.StopSequences,
            Seed = service.SeedFieldName == "seed" ? parms.Seed : null,
            RandomSeed = service.SeedFieldName == "random_seed" ? parms.Seed : null,
            ResponseFormat = parms.ResponseFormatJson ? new OpenAiResponseFormat { Type = "json_object" } : null,
            ReturnCitations = service.SupportsCitations ? parms.ReturnCitations : null,
            SearchRecencyFilter = service.SupportsSearchRecency ? parms.SearchRecency : null,
            Search = parms.SearchEnabled ? true : null
        };

        var url = ApiClient.ChatUrl(service, customBaseUrl);
        try
        {
            var (response, statusCode) = await ApiClient.Instance.RequestAsync<OpenAiRequest, OpenAiResponse>(
                url, request, service, apiKey);

            if (response.Error != null)
                return new AnalysisResponse(service, error: response.Error.Message ?? "Unknown error", httpStatusCode: statusCode);

            var content = response.Choices?.FirstOrDefault()?.Message?.Content
                ?? response.Choices?.FirstOrDefault()?.Message?.ReasoningContent;

            var usage = response.Usage;
            var tokenUsage = new TokenUsage(
                usage?.EffectiveInputTokens ?? 0,
                usage?.EffectiveOutputTokens ?? 0,
                AnalysisRepository.ExtractApiCost(usage, service));

            return new AnalysisResponse(service, analysis: content,
                error: content == null ? "No content in response" : null,
                tokenUsage: tokenUsage, httpStatusCode: statusCode)
            {
                Citations = response.Citations,
                SearchResults = response.SearchResults,
                RelatedQuestions = response.RelatedQuestions,
                RawUsageJson = Repo.FormatUsageJson(usage)
            };
        }
        catch (Exception ex)
        {
            return new AnalysisResponse(service, error: ex.Message);
        }
    }

    private static async Task<AnalysisResponse> AnalyzeWithResponsesApi(
        AppService service, string apiKey, string prompt, string model,
        AgentParameters parms, string? customBaseUrl = null)
    {
        var request = OpenAiResponsesRequest.FromString(model, prompt, parms.SystemPrompt);
        var baseUrl = customBaseUrl ?? service.BaseUrl;
        var normalizedBase = baseUrl.EndsWith("/") ? baseUrl : $"{baseUrl}/";
        var url = $"{normalizedBase}v1/responses";

        try
        {
            var (response, statusCode) = await ApiClient.Instance.RequestAsync<OpenAiResponsesRequest, OpenAiResponsesApiResponse>(
                url, request, service, apiKey);

            if (response.Error != null)
                return new AnalysisResponse(service, error: response.Error.Message ?? "Unknown error", httpStatusCode: statusCode);

            var content = response.Output?
                .Select(msg => msg.Content?.FirstOrDefault(c => c.Type == "output_text")?.Text ?? msg.Content?.FirstOrDefault()?.Text)
                .FirstOrDefault(t => t != null);

            var usage = response.Usage;
            var tokenUsage = new TokenUsage(
                usage?.EffectiveInputTokens ?? 0,
                usage?.EffectiveOutputTokens ?? 0,
                AnalysisRepository.ExtractApiCost(usage, service));

            return new AnalysisResponse(service, analysis: content,
                error: content == null ? "No content in response" : null,
                tokenUsage: tokenUsage, httpStatusCode: statusCode)
            {
                RawUsageJson = Repo.FormatUsageJson(usage)
            };
        }
        catch (Exception ex)
        {
            return new AnalysisResponse(service, error: ex.Message);
        }
    }

    public static async Task<AnalysisResponse> AnalyzeWithClaude(
        AppService service, string apiKey, string prompt, string model, AgentParameters parms)
    {
        var request = new ClaudeRequest
        {
            Model = model,
            MaxTokens = parms.MaxTokens ?? 4096,
            Messages = new List<ClaudeMessage> { new() { Role = "user", Content = prompt } },
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            System = parms.SystemPrompt,
            StopSequences = parms.StopSequences,
            Search = parms.SearchEnabled ? true : null
        };

        var baseUrl = service.BaseUrl.EndsWith("/") ? service.BaseUrl : $"{service.BaseUrl}/";
        var url = $"{baseUrl}v1/messages";

        try
        {
            var (response, statusCode) = await ApiClient.Instance.RequestAsync<ClaudeRequest, ClaudeResponse>(
                url, request, service, apiKey);

            if (response.Error != null)
                return new AnalysisResponse(service, error: response.Error.Message ?? "Unknown error", httpStatusCode: statusCode);

            var content = response.Content?.FirstOrDefault(c => c.Type == "text")?.Text;
            var tokenUsage = new TokenUsage(
                response.Usage?.InputTokens ?? 0,
                response.Usage?.OutputTokens ?? 0,
                AnalysisRepository.ExtractApiCost(response.Usage));

            return new AnalysisResponse(service, analysis: content,
                error: content == null ? "No content in response" : null,
                tokenUsage: tokenUsage, httpStatusCode: statusCode)
            {
                RawUsageJson = Repo.FormatUsageJson(response.Usage)
            };
        }
        catch (Exception ex)
        {
            return new AnalysisResponse(service, error: ex.Message);
        }
    }

    public static async Task<AnalysisResponse> AnalyzeWithGemini(
        AppService service, string apiKey, string prompt, string model, AgentParameters parms)
    {
        var config = new GeminiGenerationConfig
        {
            Temperature = parms.Temperature,
            TopP = parms.TopP,
            TopK = parms.TopK,
            MaxOutputTokens = parms.MaxTokens,
            StopSequences = parms.StopSequences
        };

        var systemInstruction = parms.SystemPrompt != null
            ? new GeminiContent { Parts = new() { new GeminiPart { Text = parms.SystemPrompt } } }
            : null;

        var request = new GeminiRequest
        {
            Contents = new() { new GeminiContent { Parts = new() { new GeminiPart { Text = prompt } }, Role = "user" } },
            GenerationConfig = config,
            SystemInstruction = systemInstruction
        };

        var url = ApiClient.GeminiGenerateUrl(service.BaseUrl, model, apiKey);

        try
        {
            var dummyService = new AppService { Id = "gemini_temp", ApiFormatStr = "GOOGLE" };
            var (response, statusCode) = await ApiClient.Instance.RequestAsync<GeminiRequest, GeminiResponse>(
                url, request, dummyService, apiKey);

            if (response.Error != null)
                return new AnalysisResponse(service, error: response.Error.Message ?? "Unknown error", httpStatusCode: statusCode);

            var content = response.Candidates?.FirstOrDefault()?.Content?.Parts?.FirstOrDefault()?.Text;
            var usage = response.UsageMetadata;
            var tokenUsage = new TokenUsage(
                usage?.PromptTokenCount ?? 0,
                usage?.CandidatesTokenCount ?? 0,
                AnalysisRepository.ExtractApiCost(usage));

            return new AnalysisResponse(service, analysis: content,
                error: content == null ? "No content in response" : null,
                tokenUsage: tokenUsage, httpStatusCode: statusCode)
            {
                RawUsageJson = Repo.FormatUsageJson(usage)
            };
        }
        catch (Exception ex)
        {
            return new AnalysisResponse(service, error: ex.Message);
        }
    }

    private static List<OpenAiMessage> BuildOpenAiMessages(string prompt, string? systemPrompt)
    {
        var messages = new List<OpenAiMessage>();
        if (!string.IsNullOrEmpty(systemPrompt))
            messages.Add(new OpenAiMessage { Role = "system", Content = systemPrompt });
        messages.Add(new OpenAiMessage { Role = "user", Content = prompt });
        return messages;
    }
}
